pymc3/backends/arviz.py

```diff
@@ -0,0 +1,606 @@
+            if shape[0] == self.nchains and shape[1] == self.ndraws:
+                data[k] = ary
+            elif shape[0] == self.nchains * self.ndraws:
+                data[k] = ary.reshape((self.nchains, self.ndraws, *shape[1:]))
+            else:
+                data[k] = np.expand_dims(ary, 0)
+                # pylint: disable=line-too-long
+                _log.warning(
+                    "posterior predictive variable %s's shape not compatible with number of chains and draws. "
+                    "This can mean that some draws or even whole chains are not represented.",
+                    k,
+                )
+        return dict_to_dataset(
+            data, library=pymc3, coords=self.coords, dims=self.dims, index_origin=self.index_origin
+        )
+
+    @requires(["posterior_predictive"])
+    def posterior_predictive_to_xarray(self):
+        """Convert posterior_predictive samples to xarray."""
+        return self.translate_posterior_predictive_dict_to_xarray(self.posterior_predictive)
+
+    @requires(["predictions"])
+    def predictions_to_xarray(self):
+        """Convert predictions (out of sample predictions) to xarray."""
+        return self.translate_posterior_predictive_dict_to_xarray(self.predictions)
+
+    def priors_to_xarray(self):
+        """Convert prior samples (and if possible prior predictive too) to xarray."""
+        if self.prior is None:
+            return {"prior": None, "prior_predictive": None}
+        if self.observations is not None:
+            prior_predictive_vars = list(self.observations.keys())
+            prior_vars = [key for key in self.prior.keys() if key not in prior_predictive_vars]
+        else:
+            prior_vars = list(self.prior.keys())
+            prior_predictive_vars = None
+
+        priors_dict = {}
+        for group, var_names in zip(
+            ("prior", "prior_predictive"), (prior_vars, prior_predictive_vars)
+        ):
+            priors_dict[group] = (
+                None
+                if var_names is None
+                else dict_to_dataset(
+                    {k: np.expand_dims(self.prior[k], 0) for k in var_names},
+                    library=pymc3,
+                    coords=self.coords,
+                    dims=self.dims,
+                    index_origin=self.index_origin,
+                )
+            )
+        return priors_dict
+
+    @requires(["observations", "multi_observations"])
+    @requires("model")
+    def observed_data_to_xarray(self):
+        """Convert observed data to xarray."""
+        if self.predictions:
+            return None
+        return dict_to_dataset(
+            {**self.observations, **self.multi_observations},
+            library=pymc3,
+            coords=self.coords,
+            dims=self.dims,
+            default_dims=[],
+            index_origin=self.index_origin,
+        )
+
+    @requires(["trace", "predictions"])
+    @requires("model")
+    def constant_data_to_xarray(self):
```

Comment:
This may also need a significant refactor, but honestly I have no idea

Reply:
Most of those attributes in `self.model` are still available, but they might correspond to different variables.  For example, `model.vars` returns the log-likelihood-space `TensorVariable`s, and the `model.*_RVs` methods return the sample-space `TensorVariable`s.

Regardless, you can always get the corresponding log-likelihood-space `TensorVariable` from the sample-space variable using `var.tag.value_var` (I should probably rename it to `var.tag.measure_var` or `var.tag.loglik_var`, though).

Reply:
All these checks and selection of variables have only the goal of gathering all named variables that are not sampled parameters nor observed/observations. Up until now that basically meant `pm.Data` objects that were not passed as `observed`, things like the `x` in a linear regression, the county and floor indexes in the radon example... Maybe we can think of a better way to identify these variables

Reply:
Absolutely; we can very easily obtain variables like that using something as simple as [`aesara.graph.basic.vars_between`](https://github.com/pymc-devs/aesara/blob/master/aesara/graph/basic.py#L781).